### PR TITLE
Query: Convert unflattened GroupJoin to correlated subquery

### LINKED ISSUE #19930: Query: Support GroupJoin when it is final query operator
This issue is about enabling following query.
```C#
context.Customers.GroupJoin(context.Orders, c => c.CustomerID, o => o.CustomerID, (c, os) => new {c, os}).ToList();
// or
from c in context.Customers
join o in context.Orders on c.CustomerID equals o.CustomerID into os
select { c, os}
```
We will not client eval GroupJoin in any of following cases
- When GroupJoin operator is followed by any other query operator
- When you apply any other operator before select in comprehension syntax (2nd query above)
- When you compose over `os` in above query in any form

All above unsupported cases will still throw client evaluation exception even after this issue is fixed.

GroupJoin operator in query which is used to generated LeftJoin in SQL as described in [our documentation](https://docs.microsoft.com/en-us/ef/core/querying/complex-query-operators#left-join) is supported and will remains so. Even after that use case, if you still need EF Core to client eval GroupJoin operator with aforementioned constraints please upvote this issue. We are not looking to client eval any other GroupJoin which appears in the middle of the query.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #19930

### CODE REVIEW

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs [hunk 1]
```diff
@@ -588,6 +601,76 @@ private Expression TryFlattenGroupJoinSelectMany(MethodCallExpression methodCall
         return methodCallExpression;
     }
 
+    private class GroupJoinConvertingExpressionVisitor : ExpressionVisitor
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:604]:
```suggestion
    private sealed class GroupJoinConvertingExpressionVisitor : ExpressionVisitor
```

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs [hunk 2]
```diff
@@ -588,6 +601,76 @@ private Expression TryFlattenGroupJoinSelectMany(MethodCallExpression methodCall
         return methodCallExpression;
     }
 
+    private sealed class GroupJoinConvertingExpressionVisitor : ExpressionVisitor
+    {
+        protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
+        {
+            if (methodCallExpression.Method.DeclaringType == typeof(Queryable)
```
Comment (roji) [inline src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:608]: Consider returning early it this isn't GroupJoin, to unindent the rest of the method body.

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs [hunk 3]
```diff
@@ -588,6 +601,76 @@ private Expression TryFlattenGroupJoinSelectMany(MethodCallExpression methodCall
...
+                    innerSource = Expression.Call(
+                        QueryableMethods.AsQueryable.MakeGenericMethod(innerSource.Type.GetSequenceType()),
+                        innerSource);
+                }
+
+                var correlationPredicate = ReplacingExpressionVisitor.Replace(
+                    outerKeySelector.Parameters[0],
+                    resultSelector.Parameters[0],
+                    Expression.AndAlso(
+                        Infrastructure.ExpressionExtensions.CreateEqualsExpression(
```
Comment (roji) [inline src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:632]: This seems to check if the outer key is null - do we support that scenario?

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs [hunk 4]
```diff
@@ -29,6 +29,19 @@ public QueryableMethodNormalizingExpressionVisitor(QueryCompilationContext query
...
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public virtual Expression Normalize(Expression expression)
+    {
+        var result = Visit(expression);
+
+        return _groupJoinConvertingExpressionVisitor.Visit(result);
```
Comment (roji) [inline src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:42]: Can we avoid double visitation here by calling into the GroupJoin conversion logic from VisitMethod (of QueryableMethodNormalizingExpressionVisitor)?

File: test/EFCore.Relational.Specification.Tests/Query/QueryNoClientEvalTestBase.cs [hunk 5]
```diff
@@ -132,14 +132,13 @@ public virtual void Throws_when_join()
     }
 
     [ConditionalFact]
-    public virtual void Throws_when_group_join()
+    public virtual void Does_not_throws_when_group_join()
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/QueryNoClientEvalTestBase.cs:135]: Can probably remove this with all the other positive tests no?

**Review by roji — APPROVED** [review]

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs (hunk 4 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:42]: We cannot. Flattening of GJ-SM requires visited children and visiting it before flattening will remove GJ.

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:632]: Join doesn't match when keys are null even if outer/inner both evaluate to null.

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs (hunk 4 above)
Comment (roji) [reply src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:42]: Ah I see, was not aware that this same visitor also does GJ flattening. We could do something fancy by havnig a pending GroupJoin bubble up and then either flatten it on SelectMany, or do the new conversion logic if not. But that may be too much.

File: src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs [hunk 6]
```diff
@@ -590,6 +603,76 @@ private Expression TryFlattenGroupJoinSelectMany(MethodCallExpression methodCall
         return methodCallExpression;
     }
 
+    private sealed class GroupJoinConvertingExpressionVisitor : ExpressionVisitor
+    {
+        protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
+        {
+            if (methodCallExpression.Method.DeclaringType == typeof(Queryable)
+                && methodCallExpression.Method.IsGenericMethod
+                && methodCallExpression.Method.GetGenericMethodDefinition()== QueryableMethods.GroupJoin)
```
Comment (roji) [inline src/EFCore/Query/Internal/QueryableMethodNormalizingExpressionVisitor.cs:612]:
```suggestion
                && methodCallExpression.Method.GetGenericMethodDefinition() == QueryableMethods.GroupJoin)
```

**Pull request merged**